### ISSUE TITLE
ERROR: tf.function, tf.data.Dataset, tensorflow-gpu incompatible  

### ISSUE THREAD

**Issue by luischinchillagarcia** [tensorflow/tensorflow#40904]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- TensorFlow version (use command below): 2.3.0 (affects all TF > 2.0.0 )
- Python version: 3.7

**Describe the current behavior**

I encountered an error when a `tf.data.Dataset` object is created or modified inside a `tf.function` graph *while* using tensorflow-gpu. Based on other issues, it seems this is caused by TF improperly placing the _VariantWrapper on a GPU. A _VariantWrapper is caused when a dataset object is created or modified inside a tf.function graph. 

Here is a small [Colab gist](https://colab.research.google.com/drive/1L_k_jomuAwKoM7RdiR-PLj7AfNZQVzh1?usp=sharing) highlighting code snippet of this issue and the related issues: [issue1](https://github.com/tensorflow/tensorflow/issues/34112), [issue2](https://github.com/tensorflow/tensorflow/issues/34519). This error has persisted since tf>2.0.0.

To put it plainly, **does this mean we should avoid wrapping any dataset operations in a tf.function**? 

eg.
```python
# Error
@tf.function
def f():
    dataset = tf.data.Dataset.from_tensor_slices(([1, 2, 3], [4, 5, 6]))
    for e in tf.range(3):
        for x, y in dataset:
            tf.print(x, y, e)
f()
```

```python
#Succeeds outside graph
dataset = tf.data.Dataset.from_tensor_slices(([1, 2, 3], [4, 5, 6]))

@tf.function
def f():
    for e in tf.range(3):
        for x, y in dataset: # `dataset` is transformed into `_VariantWrapper` rather than staying as a `tf.data.Dataset` object
            tf.print(x, y, e)
f()
```

**Describe the expected behavior**

**Standalone code to reproduce the issue**
Colab: https://colab.research.google.com/drive/1L_k_jomuAwKoM7RdiR-PLj7AfNZQVzh1?usp=sharing

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

The complete error message is: 
```python 
No unary variant device copy function found for direction: 1 and Variant type_index: 
tensorflow::data::(anonymous namespace)::DatasetVariantWrapper
```

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.3.0rc0 and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/34e39e888cee56a745c883924117498c/untitled58.ipynb).Thanks!

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.2 it encounters problem @tf.function both inside and outside the graph.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/5d2ad5300d0a5c898a6bc9e9c5462598/untitled67.ipynb).However in nightly version (`2.4.0-dev20200701`) i was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/c879985eb80eff2e8d9d60ec5325b011/untitled68.ipynb).Thanks!

**Comment by luischinchillagarcia** [comment]
Thank you for looking into it! Do you know if there is a current workaround when creating a tf.data object inside a `tf.function`?

**Comment by lindong28** [comment]
Hey @luischinchillagarcia, @ravikyram , I just noticed this issue. Sorry it has not been replied by TF team for such a long time.

The following workaround works for me on codelab. Could you try and see if it works?

```
@tf.function
def f():
    dataset = tf.data.Dataset.from_tensor_slices(([1, 2, 3], [4, 5, 6]))

    with tf.device('/CPU:0'):
        iterator = iter(dataset)

    for e in tf.range(3):
        for x, y in iterator:
            tf.print(x, y, e)
f()
```

**Comment by ravikyram** [comment]
@luischinchillagarcia @lindong28 

I have tried in colab with TF nightly version with @lindong28  suggestion and was not able to reproduce the issue even 
`dataset = tf.data.Dataset.from_tensor_slices(([1, 2, 3], [4, 5, 6]))` is wrapped inside @tf.function. Please,find the gist [here](https://colab.research.google.com/gist/ravikyram/1223e75116a1decc181fb0a3401585dc/untitled188.ipynb).

Note:
No issue with  
```
 for x, y in iterator:
            tf.print(x, y, e)
```
Still i am seeing the issue with

```
for x, y in dataset:
            tf.print(x, y, e)
```
Thanks!

**Comment by lindong28** [comment]
Thanks for the reply @ravikyram. Yes the original issue is not fixed yet and `for x, y in dataset` does not work yet.

I guess my question is, given that `for x, y in iterator`, would you be able to modify your program to replace `dataset` with `iterator` after initializing iterator in the following way?

```
with tf.device('/CPU:0'):
  iterator = iter(dataset)
```

**Comment by ravikyram** [comment]
@lindong28 

Sorry my bad .I am attaching the [gist](https://colab.research.google.com/gist/ravikyram/4a919f0d13d28ffd0f93f1fc8a469448/untitled188.ipynb) again.I am not able to reproduce the issue with your workaround.Thanks!

**Comment by ravikyram** [comment]
@luischinchillagarcia,

Please take a look at @lindong28 comment above and let us know if the issue is fixed. 
Check the [gist](https://colab.research.google.com/gist/ravikyram/4a919f0d13d28ffd0f93f1fc8a469448/untitled188.ipynb) for your reference.Thanks!

**Comment by luischinchillagarcia** [comment]
@ravikyram @lindong28 Thanks you both for aiding with this issue! It definitely helps. 

To be clear, will we be able to use tf.data within tf.function without the need to explicitly state a tf.data.Dataset as an iterator  as follows? Is this merely a temporary fix that still requires a fix down the line as @lindong28 pointed out?

```python
with tf.device('/CPU:0'):
        iterator = iter(dataset)
```

**Comment by lindong28** [comment]
Hey @luischinchillagarcia, thank you for your reply. This is just a temporary fix which hopefully can unblock your ongoing work. It is still a bug if the original program does not work. This issue needs to be fixed so that TF users would not have to make this code change.

**Comment by lindong28** [comment]
Hey everyone, the issue has been fixed in tf-nightly==2.4.0.dev20200902 and it should be fixed in the upcoming TF 2.4 release.

The fix can be validated using this colab example: https://colab.research.google.com/drive/1C4AXS8cyEsA_JmU7QVVk7UXM_uygVf05?authuser=1#scrollTo=jJBsJsRSR6AD
